File: ui/ui.check.go
package ui

import (
	"fmt"
)

func ICheckbox(name string, data ...any) *TInput {
	c := &TInput{
		as:      "checkbox",
		target:  Target(),
		name:    name,
		size:    MD,
		visible: true,
	}

	if len(data) > 0 {
		c.data = data[0]
	}

	c.Render = func(text string) string {
		if !c.visible {
			return ""
		}

		value := ""
		checked := ""

		if c.data != nil {
			// v := reflect.ValueOf(c.data)

			// if v.Kind() == reflect.Ptr {
			// 	v = v.Elem()
			// }

			// tmp := v.FieldByName(c.name)

			tmp, err := PathValue(c.data, c.name)
			if err == nil {
				value = fmt.Sprintf("%v", tmp.Interface())
			}
		}

		if value == "true" {
			checked = "checked"
		}

		c.class = Classes(c.class, "flex items-center gap-2")

		if c.required && value == "false" {
			c.class = Classes(c.class, "invalid")
		}

		return Div(Classes(c.class, c.size))(
			Input(
				Classes("cursor-pointer select-none", If(c.disabled, func() string { return DISABLED })),
				Attr{
					Value:   value,
					Checked: checked,

					Type:     c.as,
					Id:       c.target.Id,
					Name:     c.name,
					Required: c.required,
					Disabled: c.disabled,
					OnClick:  c.onclick,
					OnChange: Trim(fmt.Sprintf(`
						const el = document.getElementById('%s');

						if (el == null)
							return;

						if (el.required !== true || el.disabled === true) 
							return;

						const div = el.parentElement;

						if (el.checked) {
							div.classList.remove('invalid');
						} else {
							div.classList.add('invalid');
						}
					`, c.target.Id)),
				},
			),

			Label(&c.target).
				Required(c.required).
				ClassLabel("cursor-pointer select-none").
				Render(text),
		)
	}

	return c
}

File: ts/ui.check.ts
import { Classes, INPUT, MD } from './ui.util';
import { Div } from './ui.shared';

export class ICheckbox {
  private data?: any; private name: string; private css = ''; private size = MD;
  private required = false;
  constructor(name: string, data?: any) { this.name = name; this.data = data; }
  Class(...v: string[]) { this.css = v.join(' '); return this; }
  Size(v: string) { this.size = v; return this; }
  Required(v = true) { this.required = v; return this; }
  Render(label: string): string {
    const checked = this.data ? Boolean((this.data as any)[this.name]) : false;
    return Div(this.css)(`<label class=\"flex items-center gap-2\"><input class=\"${Classes(INPUT, this.size)}\" type=\"checkbox\" name=\"${this.name}\" ${checked ? 'checked' : ''} ${this.required ? 'required' : ''}/> ${label}</label>`);
  }
}

export const Checkbox = (name: string, data?: any) => new ICheckbox(name, data);

